engine/table/src/main/java/io/deephaven/engine/table/impl/remote/ConstructSnapshot.java

```diff
@@ -567,14 +567,30 @@ public static BarrageMessage constructBackplaneSnapshotInPositionSpace(final Obj
 
         final SnapshotFunction doSnapshot = (usePrev, beforeClockValue) -> {
             final RowSet keysToSnapshot;
-            if (positionsToSnapshot == null) {
+            if (positionsToSnapshot == null && reversePositionsToSnapshot == null) {
                 keysToSnapshot = null;
-            } else if (usePrev) {
-                try (final RowSet prevRowSet = table.getRowSet().copyPrev()) {
-                    keysToSnapshot = prevRowSet.subSetForPositions(positionsToSnapshot);
-                }
             } else {
-                keysToSnapshot = table.getRowSet().subSetForPositions(positionsToSnapshot);
+                final RowSetBuilderRandom keyBuilder = RowSetFactory.builderRandom();
+                if (usePrev) {
+                    // perform actions on the previous rowset
+                    try (final RowSet prevRowSet = table.getRowSet().copyPrev()) {
+                        if (positionsToSnapshot != null) {
+                            keyBuilder.addRowSet(prevRowSet.subSetForPositions(positionsToSnapshot));
+                        }
+                        if (reversePositionsToSnapshot != null) {
+                            keyBuilder.addRowSet(prevRowSet.subSetForReversePositions(reversePositionsToSnapshot));
+                        }
+                    }
+                } else {
+                    // perform actions on the current rowset
+                    if (positionsToSnapshot != null) {
+                        keyBuilder.addRowSet(table.getRowSet().subSetForPositions(positionsToSnapshot));
+                    }
+                    if (reversePositionsToSnapshot != null) {
+                        keyBuilder.addRowSet(table.getRowSet().subSetForReversePositions(reversePositionsToSnapshot));
+                    }
+                }
+                keysToSnapshot = keyBuilder.build();
```

Comment:
This approach is going to do more allocation than we need. It will leak the temporary row sets returned by subSetFor*Positions methods.
You want to skip the builders, and you probably want to avoid duplicating the logic.
```suggestion
                final RowSet rowSetToUse = usePrev ? table.getRowSet().copyPrev() : table.getRowSet();
                try (final SafeCloseable ignored = usePrev ? rowSetToUse : null) {
                    final WritableRowSet forwardKeys = positionsToSnapshot == null ? null :
                            rowSetToUse.subSetForPositions(positionsToSnapshot);
                    final RowSet reverseKeys = reversePositionsToSnapshot == null ? null :
                            rowSetToUse.subSetForReversePositions(reversePositionsToSnapshot);
                    if (forwardKeys != null) {
                        if (reverseKeys != null) {
                            forwardKeys.insert(reverseKeys);
                            reverseKeys.close();
                        }
                        keysToSnapshot = forwardKeys;
                    } else {
                        keysToSnapshot = reverseKeys;
                    }
                }
            }
```